apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Request.java

```diff
@@ -187,9 +191,17 @@ public CharBuffer getBodyBuffer() {
         }
     }
 
+    /**
+     * Returns the body buffer if it was written to and writing to it was finished through {@link Request#endOfBufferInput()}
+     * @return body buffer if it was written to and writing was finished; returns {@code null} otherwise.
+     */
     @Nullable
     public CharBuffer getBodyBufferForSerialization() {
```

Comment:
[question] What about returning a `CharSequence` for reading and an `Appendable` for writing instead of the `CharBuffer` itself ? Encapsulating the `CharBuffer` seems a good option here as it would prevent any call to `.flip()` when not expected.

One bonus here would be to not having to cast to `Object` in the test assertions.

Reply:
👍  Great suggestion! Will do.

Reply:
The change to `Appendable` turns out to be trickier as its users use if for decoding through `IOUtils` which relies on basic Java APIs that require `CharBuffer`, so I will only change `getBodyBufferForSerialization` for now.

Reply:
:+1: That was also something that made the rabbitmq body capture more complex than initially thought.